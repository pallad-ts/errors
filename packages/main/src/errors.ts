import {ErrorDescriptor} from "./ErrorDescriptor";
import {formatCodeFactory, PLACEHOLDER_TOKEN} from "./formatCodeFactory";


const formatCode = formatCodeFactory('E_PALLAD_ERRORS_%c');

// Needed to solve circular reference problem where ErrorDescriptor is not yet defined here
function lazyInitialization<TFactory extends (...args: any[]) => any, TDescriptorFactory extends () => ErrorDescriptor<TFactory, any>>(descriptorFactory: TDescriptorFactory) {
	let descriptor: ErrorDescriptor<any>;
	return (...args: Parameters<TFactory>) => {
		if (!descriptor) {
			descriptor = descriptorFactory();
		}
		return descriptor.create(...args);
	}
}

export const ERRORS = {
	CODE_ALREADY_TAKEN: lazyInitialization(
		() => ErrorDescriptor.useMessageFormatter(formatCode(1), (code: string) => {
			return `Error code "${code}" is already registered`;
		})
	),
	DOMAIN_IS_LOCKED: lazyInitialization(
		() => ErrorDescriptor.useDefaultMessage(formatCode(2), 'Domain is locked and no other error descriptors can be added')
	),
	ERROR_FROZEN: lazyInitialization(
		() => ErrorDescriptor.useDefaultMessage(formatCode(3), 'Cannot assign "code" property to frozen error object')
	),
	MISSING_PLACEHOLDER_IN_CODE_FORMAT: lazyInitialization(
		() => ErrorDescriptor.useMessageFormatter(formatCode(4), (format: string) => {
			return `Code placeholder "${PLACEHOLDER_TOKEN}" is missing in format "${format}"`;
		})
	),
	INVALID_ERROR_DESCRIPTOR_UNDER_KEY: lazyInitialization(
		() => ErrorDescriptor.useMessageFormatter(formatCode(5), (key: string) => {
			return `Value under key "${key}" is not valid ErrorDescriptor`;
		})
	)
};

