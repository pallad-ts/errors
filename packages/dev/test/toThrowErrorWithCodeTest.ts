import '@src/index';
import {ErrorDescriptor} from '@pallad/errors';

describe('toThrowErrorWithCode', () => {
	const descriptor = ErrorDescriptor.useDefaultMessage('C_1', 'Some default message');
	const descriptorSecond = ErrorDescriptor.useDefaultMessage('C_2', 'Some default message 2');
	const descriptorThird = ErrorDescriptor.useDefaultMessage('C_3', 'Some default message 3');

	describe('dealing with function', () => {
		it('passes if error with proper code gets thrown', () => {
			expect(() => {
				throw descriptor.create();
			})
				.toThrowErrorWithCode(descriptor);
		});

		it('passes if error with proper code gets thrown (provided as string)', () => {
			expect(() => {
				throw descriptor.create();
			})
				.toThrowErrorWithCode(descriptor.code);
		});

		it('passes if error with one of provided codes gets thrown', () => {
			expect(() => {
				throw descriptor.create();
			})
				.toThrowErrorWithCode(descriptor, descriptorSecond);

			expect(() => {
				throw descriptorSecond.create();
			})
				.toThrowErrorWithCode(descriptor, descriptorSecond);
		})

		it('fails if error with different code than provided (multiple) is thrown', () => {
			expect(() => {
				expect(() => {
					throw descriptorThird.create();
				}).toThrowErrorWithCode(descriptor, descriptorSecond);
			})
				.toThrowErrorMatchingSnapshot();
		});

		it('fails if error with different code is thrown', () => {
			expect(() => {
				expect(() => {
					throw descriptorSecond.create();
				}).toThrowErrorWithCode(descriptor);
			})
				.toThrowErrorMatchingSnapshot();
		});


		it('fails if function does not throw any error', () => {
			expect(() => {
				expect(() => {
					return 1;
				}).toThrowErrorWithCode(descriptor);
			})
				.toThrowErrorMatchingSnapshot();
		})
	});

	describe('dealing with promises', () => {
		it('passes if rejected with error with proper code', () => {
			return expect(Promise.reject(descriptor.create()))
				.rejects
				.toThrowErrorWithCode(descriptor);
		});

		it('passes if rejected with error with one of provided codes', async () => {
			await expect(Promise.reject(descriptor.create()))
				.rejects
				.toThrowErrorWithCode(descriptor, descriptorSecond);

			await expect(Promise.reject(descriptorSecond.create()))
				.rejects
				.toThrowErrorWithCode(descriptor, descriptorSecond);
		});


		it('fails if rejected with error with other code', () => {
			const result = expect(Promise.reject(descriptorSecond.create()))
				.rejects
				.toThrowErrorWithCode(descriptor);

			return expect(result)
				.rejects
				.toThrowErrorMatchingSnapshot();
		});

		it('fails if promise is not rejected', () => {
			const result = expect(Promise.resolve('test'))
				.rejects
				.toThrowErrorWithCode(descriptor);

			return expect(result).rejects.toThrowErrorMatchingSnapshot();
		})
	});

	it('fails if function or promise is not provided', () => {
		expect(() => {
			expect(1).toThrowErrorWithCode('C_1');
		})
			.toThrowErrorMatchingSnapshot();
	})
});
