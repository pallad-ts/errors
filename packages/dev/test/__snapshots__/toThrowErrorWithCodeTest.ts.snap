// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toThrowErrorWithCode dealing with function fails if error with different code is thrown 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toThrowErrorWithCode([22m[32mexpected[39m[2m)[22m

Expected error with code: C_1
Received: [31m[Error: Some default message 2][39m with code: C_2"
`;

exports[`toThrowErrorWithCode dealing with function fails if error with different code than provided (multiple) is thrown 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toThrowErrorWithCode([22m[32mexpected[39m[2m)[22m

Expected error with one of following codes: C_1, C_2
Received: [31m[Error: Some default message 3][39m with code: C_3"
`;

exports[`toThrowErrorWithCode dealing with function fails if function does not throw any error 1`] = `
"Expected the function to throw an error.
But it didn't throw anything."
`;

exports[`toThrowErrorWithCode dealing with promises fails if promise is not rejected 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mrejects[2m.[22mtoThrowErrorWithCode[2m()[22m

Received promise resolved instead of rejected
Resolved to value: [31m"test"[39m"
`;

exports[`toThrowErrorWithCode dealing with promises fails if rejected with error with other code 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toThrowErrorWithCode([22m[32mexpected[39m[2m)[22m

Expected error with code: C_1
Received: [31m[Error: Some default message 2][39m with code: C_2"
`;

exports[`toThrowErrorWithCode fails if function or promise is not provided 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).toThrowErrorWithCode([22m[32mtype[39m[2m, [22m[32mmessage[39m[2m)[22m

Received value must be a function but instead "1" was found"
`;
